runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -573,12 +574,331 @@ static TR::Instruction *generatePrefetchAfterHeaderAccess(TR::Node
+
+         fpcw = comp->getJittedMethodSymbol()->usesSinglePrecisionMode() ?
+                SINGLE_PRECISION_ROUND_TO_NEAREST : DOUBLE_PRECISION_ROUND_TO_NEAREST;
+
+         generateMemInstruction(LDCWMem, node, generateX86MemoryReference(cg->findOrCreate2ByteConstant(node, fpcw), cg), cg);
+         }
+
+      // WARNING:
+      //
+      // The following load instructions are dissected in the snippet to determine the target registers.
+      // If they or their format is changed, you may need to change the snippet also.
+      //
+      loadHighInstr = generateRegMemInstruction(L4RegMem, node, highReg,
+                                                generateX86MemoryReference(*convertedLongMR, 4, cg), cg);
+
+      loadLowInstr = generateRegMemInstruction(L4RegMem, node, lowReg,
+                                               generateX86MemoryReference(*convertedLongMR, 0, cg), cg);
+
+      // Jump to the snippet if the converted value is an indefinite integer; otherwise continue.
+      //
+      generateRegImmInstruction(CMP4RegImm4, node, highReg, INT_MIN, cg);
+      generateLabelInstruction(JNE4, node, reStartLabel, cg);
+      generateRegRegInstruction(TEST4RegReg, node, lowReg, lowReg, cg);
+      generateLabelInstruction(JE4, node, snippetLabel, cg);
+
+      // Create the conversion snippet.
+      //
+      cg->addSnippet( new (cg->trHeapMemory()) TR::X86FPConvertToLongSnippet(reStartLabel,
+                                                                             snippetLabel,
+                                                                             helperSymRef,
+                                                                             node,
+                                                                             loadHighInstr,
+                                                                             loadLowInstr,
+                                                                             cg) );
+
+      // Make sure the high and low long registers are assigned to something.
+      //
+      if (accReg)
+         {
+         deps = generateRegisterDependencyConditions((uint8_t)0, 3, cg);
+         deps->addPostCondition(accReg, TR::RealRegister::eax, cg);
+         deps->addPostCondition(lowReg, TR::RealRegister::NoReg, cg);
+         deps->addPostCondition(highReg, TR::RealRegister::NoReg, cg);
+         }
+      else
+         {
+         deps = generateRegisterDependencyConditions((uint8_t)0, 2, cg);
+         deps->addPostCondition(lowReg, TR::RealRegister::NoReg, cg);
+         deps->addPostCondition(highReg, TR::RealRegister::NoReg, cg);
+         }
+      generateLabelInstruction(LABEL, node, reStartLabel, deps, cg);
+
+      // We want the floating point register to be live through the snippet, so if it is
+      // not referenced again we must pop it off the stack here.
+      //
+      if ((cg->decReferenceCount(child) == 0) || (child->getRegister()->getKind() == TR_FPR))
```

Comment:
Isn't this conditional always true now and can be folded away?  I believe the child's kind (since it is either float or double) will always be a `TR_FPR`.

Reply:
Seems to be reachable on OMR tests

Reply:
Sorry, I don't quite understand your comment.  Are you saying you agree with me, or there are circumstances where a non-FPR (i.e., x87 register) reaches here?  If it is the latter, I'd like to understand the situation in which that occurs.

Reply:
Yes, it seems that the child can be an x87 register. I'll look into it.

Reply:
Nevermind, sorry, I misread that